### PR TITLE
Feature/#49 observable mount

### LINKED ISSUE #49: Observe status of mounted filesystem
A dokany mount can be unmounted by external events. We should give downstream libraries the possibility to react to such events by either returning an observable/registerable object when the `mount()` method is called and let them register an observer.

## Remark

This is an upstream issue for https://github.com/cryptomator/cryptomator/issues/1508

### PR DESCRIPTION
**Pull request by infeo**
Closes #49 

This PR adds the capability to register a Runnable instance during `mount(...)`, which is executed when the mount is either unmounted or the underlying process is terminated.

Additionally, some refactoring regarding exceptions is done.

### CODE REVIEW

File: src/main/java/org/cryptomator/frontend/dokany/MountFactory.java [hunk 1]
```diff
@@ -103,11 +103,11 @@ public Mount mount(Path fileSystemRoot, Path mountPoint, String volumeName, Stri
 	 * @param volumeName The name of the drive as shown to the user.
 	 * @param fileSystemName The technical file system name shown in the drive properties window.
 	 * @param additionalOptions Additional options for the mount. For any unset option a default value is used. See {@link MountUtil} for details.
-	 * @param afterUnmountAction action to be performed after the filesystem is unmounted.
+	 * @param afterMountExitAction action to be performed after the dokan mount exited (regularly or irregularly)
 	 * @return The mount object.
 	 * @throws DokanyMountFailedException if the mount process is aborted due to errors
 	 */
-	public Mount mount(Path fileSystemRoot, Path mountPoint, String volumeName, String fileSystemName, String additionalOptions, Runnable afterUnmountAction) throws DokanyMountFailedException {
+	public Mount mount(Path fileSystemRoot, Path mountPoint, String volumeName, String fileSystemName, String additionalOptions, Runnable afterMountExitAction) throws DokanyMountFailedException {
```
Comment (overheadhunter) [inline src/main/java/org/cryptomator/frontend/dokany/MountFactory.java:110]: I'm still not 100% happy with the name `afterMountExitAction`. The fact that it contains "mount" might still be misleading. Maybe the shorter `onExit` is generic enough to not create expectations about the timing, so you're forced to rtfm.

Comment (infeo) [reply src/main/java/org/cryptomator/frontend/dokany/MountFactory.java]: I renamed it to `onDokanExit` in 8045576. It is more specifc then "onExit", but does not contain anything about the mounting in its name.

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 2]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
 				//real mount op
-				mountFuture = CompletableFuture
-						.supplyAsync(() -> NativeMethods.DokanMain(deviceOptions, new DokanyOperationsProxy(fileSystem)), executor)
-						.handle((returnVal, exception) -> {
-							setIsMounted(false);
-							if (returnVal != null && returnVal.equals(0)) {
-								return 0;
+				CompletableFuture.supplyAsync(() -> NativeMethods.DokanMain(deviceOptions, new DokanyOperationsProxy(fileSystem)), executor)
+						.whenComplete((returnVal, throwable) -> {
+							isMounted = false;
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:107]: changes volatile `isMounted` without synchronized block

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 3]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
-					if (e.getCause() instanceof DokanyException) {
-						throw (DokanyException) e.getCause();
-					} else {
-						throw new DokanyException(e);
-					}
-				} catch (InterruptedException e) {
-					Thread.currentThread().interrupt();
-				}
+						})
+						.get(1000, TimeUnit.MILLISECONDS);
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:116]: timeout seems to be kind of arbitrary. maybe no need for customization, but at least make it a constant

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 4]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
+							onDokanExit.run();
+							if (throwable != null) {
+								throw new DokanyException(throwable);
 							}
-
-							if (returnVal == null) {
-								throw new DokanyException(exception);
-							} else {
-								throw new DokanyException("Return Code " + returnVal + ": " + MountError.fromInt(returnVal).getDescription());
+							if (returnVal != null && returnVal != 0) {
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:112]: Can we `assert returnVal != null`? At this point either throwable or returnVal should be non-null (if I'm not mistaken).

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 5]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
-					}
-				} catch (InterruptedException e) {
-					Thread.currentThread().interrupt();
-				}
+						})
+						.get(1000, TimeUnit.MILLISECONDS);
 
-				setIsMounted(true);
+				//if the execution reaches this point, device was directly unmounted.
+				throw new DokanyException("Mount failed: Volume was instantly unmounted.");
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:119]: Tbh, I don't understand what should prevent the program from reaching this point.

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 6]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
 			} catch (UnsatisfiedLinkError err) {
-				LOG.error("Unable to load dokan driver.", err);
 				throw new DokanyException(err);
+			} catch (ExecutionException e) {
+				if (e.getCause() instanceof DokanyException) {
+					throw (DokanyException) e.getCause();
+				} else {
+					throw new DokanyException(e.getCause());
+				}
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:127]: Do we use Guava in this lib? If yes, we can simplify this to

```suggestion
				throwIfInstanceOf(e.getCause(), DokanyException.class);
				throw new DokanyException(e.getCause());
```

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 7]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+			} catch (TimeoutException e) {
+				//up and running
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:131]: ??

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 8]
```diff
@@ -64,70 +62,79 @@ public long getVersion() {
...
-				LOG.error("Unable to load dokan driver.", err);
 				throw new DokanyException(err);
+			} catch (ExecutionException e) {
+				if (e.getCause() instanceof DokanyException) {
+					throw (DokanyException) e.getCause();
+				} else {
+					throw new DokanyException(e.getCause());
+				}
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:129]:
```suggestion
				Thread.currentThread().interrupt();
				throw new DokanyException("Dokany mount interrupted.");
```

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 9]
```diff
@@ -170,7 +177,11 @@ public void unmountForced() {
 
 	@Override
 	public void reveal(Revealer revealer) throws Exception {
-		revealer.reveal(Path.of(deviceOptions.MountPoint.toString()));
+		if (isMounted) {
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:180]: access to volatile `isMounted` without synchronized block

**Review by overheadhunter — CHANGES_REQUESTED** [review]
It seems like there are timing side effects used for branching

File: src/main/java/com/dokany/java/DokanyMount.java [hunk 10]
```diff
@@ -9,25 +9,27 @@
...
+import java.util.function.Consumer;
 
 /**
  * Main class to start and stop Dokany file system.
  */
 public final class DokanyMount implements Mount {
 
 	private static final Logger LOG = LoggerFactory.getLogger(DokanyMount.class);
+	private static final int MOUNT_TIMEOUT_MILLIS = 3000;
+	private static final AtomicInteger mountCounter = new AtomicInteger(1);
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyMount.java:26]: `MOUNT_COUNTER`

File: src/main/java/com/dokany/java/DokanyOperationsProxy.java [hunk 11]
```diff
@@ -121,8 +126,8 @@
 		public long callback(WString rawPath, WinBase.SECURITY_ATTRIBUTES securityContext, int rawDesiredAccess, int rawFileAttributes, int rawShareAccess, int rawCreateDisposition, int rawCreateOptions, DokanyFileInfo dokanyFileInfo) {
 			try {
 				//hack to uniquely identify the dokan threads
-				if (Thread.currentThread().getName().equals(DEFAULT_THREAD_NAME)) {
-					Thread.currentThread().setName("dokan-" + threadCounter.getAndIncrement());
+				if (Thread.currentThread().getName().equals(nativeThreadNamePrefix)) {
+					Thread.currentThread().setName(nativeThreadNamePrefix + threadCounter.getAndIncrement());
```
Comment (overheadhunter) [inline src/main/java/com/dokany/java/DokanyOperationsProxy.java:130]: Where is this Thread created? Can we change the ThreadFactory in order to set the name a priori?

**Review by overheadhunter — CHANGES_REQUESTED** [review]

File: src/main/java/com/dokany/java/DokanyOperationsProxy.java (hunk 11 above)
Comment (infeo) [reply src/main/java/com/dokany/java/DokanyOperationsProxy.java]: > Where is this Thread created? 

Those threads are the native callback threads of Dokany. I removed this hack and implemented our own CallbackThreadInitializer (see f02686b)

File: src/main/java/com/dokany/java/DokanyMount.java (hunk 10 above)
Comment (infeo) [reply src/main/java/com/dokany/java/DokanyMount.java]: fixed in 4dac384

**Pull request merged**